File: realsense_ros2/src/python/vio_to_px4.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from px4_msgs.msg import VehicleOdometry
from nav_msgs.msg import Odometry
import numpy as np
from time import time

class VIOPublisher(Node):
    def __init__(self):
        super().__init__('vio_publisher')
        self.odom_sub = self.create_subscription(Odometry, '/odom', self.odom_callback, 10)
        self.px4_pub = self.create_publisher(VehicleOdometry, '/fmu/in/vehicle_visual_odometry', 10)
        
        self.is_publishing = False
        # Keep track of the last message received time
        self.last_odom_time = time()  
        self.odom_timeout = 5.0  # Timeout threshold in seconds
        
        # Timer to check subscription status every 1 second
        self.timer = self.create_timer(1.0, self.check_odom_status)

    def odom_callback(self, msg):
        # directly forwarding
        if not self.is_publishing:
            self.get_logger().info('/odom will be forwareded to /fmu/in/vehicle_odometry')
            self.is_publishing = True
        
        # Update last received time
        self.last_odom_time = time()
        
        vio_msg = VehicleOdometry()
        vio_msg.timestamp = int(self.get_clock().now().nanoseconds / 1000)  # PX4 expects timestamps in microseconds
        vio_msg.position = [msg.pose.pose.position.x, msg.pose.pose.position.y, msg.pose.pose.position.z]

        # Convert quaternion to yaw
        q = msg.pose.pose.orientation
        yaw = np.arctan2(2.0 * (q.w * q.z + q.x * q.y), 1.0 - 2.0 * (q.y * q.y + q.z * q.z))
        vio_msg.q = [q.x, q.y, q.z, q.w]  # PX4 uses [x, y, z, w] order

        # Set velocity (if available)
        vio_msg.velocity = [msg.twist.twist.linear.x, msg.twist.twist.linear.y, msg.twist.twist.linear.z]
        vio_msg.angular_velocity = [msg.twist.twist.angular.x, msg.twist.twist.angular.y, msg.twist.twist.angular.z]

        vio_msg.velocity_frame = VehicleOdometry.VELOCITY_FRAME_NED  # PX4 uses NED frame

        self.px4_pub.publish(vio_msg)
    
    def check_odom_status(self):
        # Check if /odom topic is still alive by checking the last received time
        if time() - self.last_odom_time > self.odom_timeout and self.is_publishing:
            self.get_logger().warning('/odom topic has not published for {:.2f} seconds.'.format(time() - self.last_odom_time))
            self.is_publishing = False
        else:
            pass

def main():
    rclpy.init()
    node = VIOPublisher()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: realsense_ros2/src/vio_to_px4.cpp
#include "rclcpp/rclcpp.hpp"
#include "nav_msgs/msg/odometry.hpp"

class MinimalSubscriber : public rclcpp::Node
{
public:
  MinimalSubscriber()
  : Node("minimal_subscriber")
  {
    // Create a subscription to the "/camera/pose/sample" topic
    subscription_ = this->create_subscription<nav_msgs::msg::Odometry>(
      "/camera/pose/sample", 10, std::bind(&MinimalSubscriber::listener_callback, this, std::placeholders::_1));
  }

private:
  // Callback function for the subscription
  void listener_callback(const nav_msgs::msg::Odometry::SharedPtr msg)
  {
    RCLCPP_INFO(this->get_logger(), "Received a message on /camera/pose/sample");
  }

  // Subscription object
  rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr subscription_;
};

int main(int argc, char * argv[])
{
  // Initialize ROS 2
  rclcpp::init(argc, argv);

  // Create an instance of the MinimalSubscriber node
  auto node = std::make_shared<MinimalSubscriber>();
  RCLCPP_INFO(node->get_logger(), "node initialized");
  // Spin the node to process incoming messages
  rclcpp::spin(node);

  // Shutdown ROS 2
  rclcpp::shutdown();
  return 0;
}
